### DISCUSSION TITLE
use `impl trait` in grammar paramater

### DISCUSSION THREAD

**Discussion by AphrosDelos532** [lalrpop/lalrpop#872] [General]
I have a question about a simplified version of my project. Below is the project structure.

```txt
├── Cargo.lock
├── Cargo.toml
└── src
    ├── grammar.lalrpop
    ├── main.rs
    └── traits.rs
```

The `traits.rs` file contains the following content. I have defined and exposed a `trait` named `AccessCache`.

```rs
struct Cache {}

pub trait AccessCache {}

impl AccessCache for Cache {}

pub fn fun(cache: &impl AccessCache) {}
```

However, when I attempt to add it to the `parser`, I encounter the following error:

```lalrpop
use crate::traits::AccessCache;

grammar(pool: &impl AccessCache);

...
```

```txt
Caused by:
  process didn't exit successfully: `.../build-script-build` (exit status: 1)
  
  --- stdout
  processing file `src/parser/parser.lalrpop`
  src/parser/parser.lalrpop:17:21: 17:31 error: unexpected token: `AccessCache`

  --- stderr
    grammar(pool: &impl AccessCache);
                        ~~~~~~~~~~~
```

However, I can use it well in other Rust functions. I'm puzzled. Is it not possible to use it in this way?

**Comment by Pat-Lafon** [comment]
Yes, unfortunately grammar parameters are incomplete in their handling of more advanced rust functionality. See this similar issue: https://github.com/lalrpop/lalrpop/issues/134 Please open an issue and we would be very interested in pr's on this.
